File: app/src/main/java/com/bitflaker/lucidsourcekit/data/datastore/DataStoreKeys.java
package com.bitflaker.lucidsourcekit.data.datastore;

import androidx.datastore.preferences.core.Preferences;
import androidx.datastore.preferences.core.PreferencesKeys;

import java.util.HashMap;

public class DataStoreKeys {
    // Preferences
    public static final Preferences.Key<String> LANGUAGE = PreferencesKeys.stringKey("LANGUAGE");

    // App open streak
    public static final Preferences.Key<Long> FIRST_OPEN_TIME_TODAY_DAY = PreferencesKeys.longKey("FIRST_OPEN_TIME_TODAY_DAY");
    public static final Preferences.Key<Long> FIRST_OPEN_LATEST_DAY = PreferencesKeys.longKey("FIRST_OPEN_LATEST_DAY");
    public static final Preferences.Key<Long> APP_OPEN_STREAK = PreferencesKeys.longKey("APP_OPEN_STREAK");
    public static final Preferences.Key<Long> APP_OPEN_STREAK_LONGEST = PreferencesKeys.longKey("APP_OPEN_STREAK_LONGEST");

    // Goals settings
    public static final Preferences.Key<Boolean> GOAL_DIFFICULTY_AUTO_ADJUST = PreferencesKeys.booleanKey("GOAL_DIFFICULTY_AUTO_ADJUST");
    public static final Preferences.Key<Float> GOAL_DIFFICULTY_VALUE_EASY = PreferencesKeys.floatKey("GOAL_DIFFICULTY_VALUE_EASY");
    public static final Preferences.Key<Float> GOAL_DIFFICULTY_VALUE_NORMAL = PreferencesKeys.floatKey("GOAL_DIFFICULTY_VALUE_NORMAL");
    public static final Preferences.Key<Float> GOAL_DIFFICULTY_VALUE_HARD = PreferencesKeys.floatKey("GOAL_DIFFICULTY_VALUE_HARD");
    public static final Preferences.Key<Float> GOAL_DIFFICULTY_VALUE_COMMON = PreferencesKeys.floatKey("GOAL_DIFFICULTY_VALUE_COMMON");
    public static final Preferences.Key<Float> GOAL_DIFFICULTY_VALUE_UNCOMMON = PreferencesKeys.floatKey("GOAL_DIFFICULTY_VALUE_UNCOMMON");
    public static final Preferences.Key<Float> GOAL_DIFFICULTY_VALUE_RARE = PreferencesKeys.floatKey("GOAL_DIFFICULTY_VALUE_RARE");
    public static final Preferences.Key<Float> GOAL_DIFFICULTY_VALUE_VARIANCE = PreferencesKeys.floatKey("GOAL_DIFFICULTY_VALUE_VARIANCE");
    public static final Preferences.Key<Float> GOAL_DIFFICULTY_VARIANCE = PreferencesKeys.floatKey("GOAL_DIFFICULTY_VARIANCE");
    public static final Preferences.Key<Float> GOAL_DIFFICULTY_TENDENCY = PreferencesKeys.floatKey("GOAL_DIFFICULTY_TENDENCY");
    public static final Preferences.Key<Integer> GOAL_DIFFICULTY_ACCURACY = PreferencesKeys.intKey("GOAL_DIFFICULTY_ACCURACY");
    public static final Preferences.Key<Integer> GOAL_DIFFICULTY_COUNT = PreferencesKeys.intKey("GOAL_DIFFICULTY_COUNT");
    public static final Preferences.Key<Float> GOAL_FUNCTION_VALUE_A = PreferencesKeys.floatKey("GOAL_FUNCTION_VALUE_A");
    public static final Preferences.Key<Float> GOAL_FUNCTION_VALUE_B = PreferencesKeys.floatKey("GOAL_FUNCTION_VALUE_B");
    public static final Preferences.Key<Float> GOAL_FUNCTION_VALUE_C = PreferencesKeys.floatKey("GOAL_FUNCTION_VALUE_C");

    // Authentication
    public static final Preferences.Key<String> AUTHENTICATION_TYPE = PreferencesKeys.stringKey("AUTHENTICATION_TYPE");
    public static final Preferences.Key<String> AUTHENTICATION_HASH = PreferencesKeys.stringKey("AUTHENTICATION_HASH");
    public static final Preferences.Key<String> AUTHENTICATION_SALT = PreferencesKeys.stringKey("AUTHENTICATION_SALT");
    public static final Preferences.Key<Boolean> AUTHENTICATION_USE_BIOMETRICS = PreferencesKeys.booleanKey("AUTHENTICATION_USE_BIOMETRICS");

    // App setup
    public static final Preferences.Key<Boolean> APP_SETUP_FINISHED = PreferencesKeys.booleanKey("APP_SETUP_FINISHED");

    // Notifications
    public static final Preferences.Key<Boolean> NOTIFICATION_PAUSED_ALL = PreferencesKeys.booleanKey("NOTIFICATION_PAUSED_ALL");
    public static final Preferences.Key<String> NOTIFICATION_NEXT_CATEGORY = PreferencesKeys.stringKey("NOTIFICATION_NEXT_CATEGORY");
    public static final Preferences.Key<Boolean> NOTIFICATION_RC_REMINDER_FULL_SCREEN = PreferencesKeys.booleanKey("NOTIFICATION_RC_REMINDER_FULL_SCREEN");
    public static final Preferences.Key<byte[]> NOTIFICATION_RC_REMINDER_FULL_SCREEN_CONFIRM_DIGITS = PreferencesKeys.byteArrayKey("NOTIFICATION_RC_REMINDER_FULL_SCREEN_CONFIRM_DIGITS");
    public static final Preferences.Key<Long> NOTIFICATION_RC_REMINDER_FULL_SCREEN_CONFIRM_TIME = PreferencesKeys.longKey("NOTIFICATION_RC_REMINDER_FULL_SCREEN_CONFIRM_TIME");

    // Default settings values
    public static final HashMap<Preferences.Key<?>, Object> DEFAULT_VALUES = new HashMap<>() {{
        this.put(LANGUAGE, "en");
        this.put(FIRST_OPEN_LATEST_DAY, 0L);
        this.put(APP_OPEN_STREAK, 0L);
        this.put(APP_OPEN_STREAK_LONGEST, 0L);
        this.put(GOAL_DIFFICULTY_AUTO_ADJUST, true);
        this.put(GOAL_DIFFICULTY_VALUE_EASY, 100.0f);
        this.put(GOAL_DIFFICULTY_VALUE_NORMAL, 100.0f);
        this.put(GOAL_DIFFICULTY_VALUE_HARD, 100.0f);
        this.put(GOAL_DIFFICULTY_VALUE_COMMON, 0.3333333f);
        this.put(GOAL_DIFFICULTY_VALUE_UNCOMMON, 0.3333333f);
        this.put(GOAL_DIFFICULTY_VALUE_RARE, 0.3333333f);
        this.put(GOAL_DIFFICULTY_VALUE_VARIANCE, 10.0f);
        this.put(GOAL_DIFFICULTY_VARIANCE, 0.15f);
        this.put(GOAL_DIFFICULTY_TENDENCY, 1.8f);
        this.put(GOAL_DIFFICULTY_ACCURACY, 100);
        this.put(GOAL_DIFFICULTY_COUNT, 3);
        this.put(GOAL_FUNCTION_VALUE_A, 0.0f);
        this.put(GOAL_FUNCTION_VALUE_B, 0.0f);
        this.put(GOAL_FUNCTION_VALUE_C, 100.0f);
        this.put(AUTHENTICATION_TYPE, "none");
        this.put(AUTHENTICATION_HASH, "");
        this.put(AUTHENTICATION_SALT, "");
        this.put(AUTHENTICATION_USE_BIOMETRICS, false);
        this.put(NOTIFICATION_PAUSED_ALL, false);
        this.put(NOTIFICATION_NEXT_CATEGORY, "NONE");
        this.put(NOTIFICATION_RC_REMINDER_FULL_SCREEN, false);
        this.put(NOTIFICATION_RC_REMINDER_FULL_SCREEN_CONFIRM_DIGITS, new byte[0]);
        this.put(NOTIFICATION_RC_REMINDER_FULL_SCREEN_CONFIRM_TIME, 0L);
        this.put(APP_SETUP_FINISHED, false);
    }};
}

File: app/src/main/java/com/bitflaker/lucidsourcekit/data/datastore/DataStoreKeys.kt
package com.bitflaker.lucidsourcekit.data.datastore

import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.booleanPreferencesKey
import androidx.datastore.preferences.core.byteArrayPreferencesKey
import androidx.datastore.preferences.core.floatPreferencesKey
import androidx.datastore.preferences.core.intPreferencesKey
import androidx.datastore.preferences.core.longPreferencesKey
import androidx.datastore.preferences.core.stringPreferencesKey
import com.bitflaker.lucidsourcekit.utils.Tools
import java.util.Calendar

object DataStoreKeys {
    // Preferences
    val LANGUAGE = DefaultPreferenceKey("LANGUAGE", "en")

    // App open streak
    val FIRST_OPEN_TIME_TODAY_DAY = DefaultPreferenceKey("FIRST_OPEN_TIME_TODAY_DAY", Calendar.getInstance().timeInMillis)
    val FIRST_OPEN_LATEST_DAY = DefaultPreferenceKey("FIRST_OPEN_LATEST_DAY", Tools.getMidnightTime())
    val APP_OPEN_STREAK = DefaultPreferenceKey("APP_OPEN_STREAK", 0L)
    val APP_OPEN_STREAK_LONGEST = DefaultPreferenceKey("APP_OPEN_STREAK_LONGEST", 0L)

    // Goals settings
    val GOAL_DIFFICULTY_AUTO_ADJUST = DefaultPreferenceKey("GOAL_DIFFICULTY_AUTO_ADJUST", true)
    val GOAL_DIFFICULTY_VALUE_EASY = DefaultPreferenceKey("GOAL_DIFFICULTY_VALUE_EASY", 100.0f)
    val GOAL_DIFFICULTY_VALUE_NORMAL = DefaultPreferenceKey("GOAL_DIFFICULTY_VALUE_NORMAL", 100.0f)
    val GOAL_DIFFICULTY_VALUE_HARD = DefaultPreferenceKey("GOAL_DIFFICULTY_VALUE_HARD", 100.0f)
    val GOAL_DIFFICULTY_VALUE_COMMON = DefaultPreferenceKey("GOAL_DIFFICULTY_VALUE_COMMON", 0.3333333f)
    val GOAL_DIFFICULTY_VALUE_UNCOMMON = DefaultPreferenceKey("GOAL_DIFFICULTY_VALUE_UNCOMMON", 0.3333333f)
    val GOAL_DIFFICULTY_VALUE_RARE = DefaultPreferenceKey("GOAL_DIFFICULTY_VALUE_RARE", 0.3333333f)
    val GOAL_DIFFICULTY_VALUE_VARIANCE = DefaultPreferenceKey("GOAL_DIFFICULTY_VALUE_VARIANCE", 10.0f)
    val GOAL_DIFFICULTY_VARIANCE = DefaultPreferenceKey("GOAL_DIFFICULTY_VARIANCE", 0.15f)
    val GOAL_DIFFICULTY_TENDENCY = DefaultPreferenceKey("GOAL_DIFFICULTY_TENDENCY", 1.8f)
    val GOAL_DIFFICULTY_ACCURACY = DefaultPreferenceKey("GOAL_DIFFICULTY_ACCURACY", 100)
    val GOAL_DIFFICULTY_COUNT = DefaultPreferenceKey("GOAL_DIFFICULTY_COUNT", 3)
    val GOAL_FUNCTION_VALUE_A = DefaultPreferenceKey("GOAL_FUNCTION_VALUE_A", 0.0f)
    val GOAL_FUNCTION_VALUE_B = DefaultPreferenceKey("GOAL_FUNCTION_VALUE_B", 0.0f)
    val GOAL_FUNCTION_VALUE_C = DefaultPreferenceKey("GOAL_FUNCTION_VALUE_C", 100.0f)

    // Authentication
    val AUTHENTICATION_TYPE = DefaultPreferenceKey("AUTHENTICATION_TYPE", "none")
    val AUTHENTICATION_HASH = DefaultPreferenceKey("AUTHENTICATION_HASH", "")
    val AUTHENTICATION_SALT = DefaultPreferenceKey("AUTHENTICATION_SALT", "")
    val AUTHENTICATION_USE_BIOMETRICS = DefaultPreferenceKey("AUTHENTICATION_USE_BIOMETRICS", false)

    // App setup
    val APP_SETUP_FINISHED = DefaultPreferenceKey("APP_SETUP_FINISHED", false)

    // Notifications
    val NOTIFICATION_PAUSED_ALL = DefaultPreferenceKey("NOTIFICATION_PAUSED_ALL", false)
    val NOTIFICATION_NEXT_CATEGORY = DefaultPreferenceKey("NOTIFICATION_NEXT_CATEGORY", "NONE")
    val NOTIFICATION_RC_REMINDER_FULL_SCREEN = DefaultPreferenceKey("NOTIFICATION_RC_REMINDER_FULL_SCREEN", false)
    val NOTIFICATION_RC_REMINDER_FULL_SCREEN_CONFIRM_DIGITS = DefaultPreferenceKey("NOTIFICATION_RC_REMINDER_FULL_SCREEN_CONFIRM_DIGITS", ByteArray(0))
    val NOTIFICATION_RC_REMINDER_FULL_SCREEN_CONFIRM_TIME = DefaultPreferenceKey("NOTIFICATION_RC_REMINDER_FULL_SCREEN_CONFIRM_TIME", 0L)

    // Usage stats
    val USAGE_STATS_PERMISSION_DISMISSED = DefaultPreferenceKey("USAGE_STATS_PERMISSION_DISMISSED", 0)
}

class DefaultPreferenceKey<T>(val name: String, val default: T & Any) {
    val preferenceKey: Preferences.Key<T> by lazy {
        createPreferenceKey(default)
    }

    /**
     * Creates a [Preferences.Key] with the name provided in the constructor and of the
     * same type as the default value. This is a workaround for the inability to generically
     * create the keys from the actual type
     * @param defaultValue The default value to be used in case the parameter is not initialized
     * @param T The type of the key. This type has to be supported by [Preferences.Key]
     * @return The constructed preference key of the type provided by the default value
     */
    fun <T> createPreferenceKey(defaultValue: T & Any): Preferences.Key<T> = when (defaultValue::class) {
        String::class -> stringPreferencesKey(name)
        Int::class -> intPreferencesKey(name)
        Long::class -> longPreferencesKey(name)
        ByteArray::class -> byteArrayPreferencesKey(name)
        Boolean::class -> booleanPreferencesKey(name)
        Float::class -> floatPreferencesKey(name)
        else -> throw IllegalArgumentException("Unsupported default value preference")
    } as Preferences.Key<T>
}